model/histogram/float_histogram_test.go

```diff
@@ -2252,3 +2252,520 @@ func TestFloatBucketIteratorTargetSchema(t *testing.T) {
+				NegativeSpans:   []Span{{4, 2}, {1, 2}},
+				NegativeBuckets: []float64{1, 1, 4, 4},
+			},
+			&FloatHistogram{
+				ZeroThreshold:   0.25,
+				ZeroCount:       12,
+				Count:           30,
+				Sum:             2.345,
+				Schema:          1,
+				PositiveSpans:   []Span{{-3, 2}, {5, 5}},
+				PositiveBuckets: []float64{1, 0, 3, 2, 2, 3, 4},
+				NegativeSpans:   []Span{{6, 3}, {6, 4}},
+				NegativeBuckets: []float64{3, 0.5, 0.5, 2, 3, 2, 4},
+			},
+			&FloatHistogram{
+				ZeroThreshold:   0.25,
+				ZeroCount:       22,
+				Count:           51,
+				Sum:             3.579,
+				PositiveSpans:   []Span{{-1, 7}},
+				PositiveBuckets: []float64{6, 4, 2, 6, 10, 9, 5},
+				NegativeSpans:   []Span{{3, 3}, {1, 3}},
+				NegativeBuckets: []float64{3, 2, 1, 4, 9, 6},
+			},
+		},
+		{
+			"schema change combined with larger zero bucket in first histogram",
+			&FloatHistogram{
+				ZeroThreshold:   0.25,
+				ZeroCount:       8,
+				Count:           21,
+				Sum:             1.234,
+				Schema:          0,
+				PositiveSpans:   []Span{{-1, 4}, {0, 3}},
+				PositiveBuckets: []float64{5, 4, 2, 3, 6, 2, 5},
+				NegativeSpans:   []Span{{4, 2}, {1, 2}},
+				NegativeBuckets: []float64{1, 1, 4, 4},
+			},
+			&FloatHistogram{
+				ZeroThreshold:   0.01,
+				ZeroCount:       11,
+				Count:           30,
+				Sum:             2.345,
+				Schema:          1,
+				PositiveSpans:   []Span{{-4, 3}, {5, 5}},
+				PositiveBuckets: []float64{1, 0, 0, 3, 2, 2, 3, 4},
+				NegativeSpans:   []Span{{6, 3}, {6, 4}},
+				NegativeBuckets: []float64{3, 0.5, 0.5, 2, 3, 2, 4},
+			},
+			&FloatHistogram{
+				ZeroThreshold:   0.25,
+				ZeroCount:       20,
+				Count:           51,
+				Sum:             3.579,
+				PositiveSpans:   []Span{{-1, 4}, {0, 3}},
+				PositiveBuckets: []float64{5, 4, 2, 6, 10, 9, 5},
+				NegativeSpans:   []Span{{3, 3}, {1, 3}},
+				NegativeBuckets: []float64{3, 2, 1, 4, 9, 6},
+			},
+		},
+	}
+
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			require.Equal(t, c.expected, c.in1.AddNew(c.in2))
+			// Has it also happened in-place?
+			require.Equal(t, c.expected, c.in1)
+		})
+	}
+}
+
+func BenchmarkAddOld(b *testing.B) {
+	// run the Fib function b.N times
+
+	f1 := &FloatHistogram{
+		ZeroThreshold:   0.01,
+		ZeroCount:       8,
+		Count:           21,
+		Sum:             1.234,
+		Schema:          0,
+		PositiveSpans:   []Span{{-1, 4}, {0, 3}},
+		PositiveBuckets: []float64{5, 4, 2, 3, 6, 2, 5},
+		NegativeSpans:   []Span{{4, 2}, {1, 2}},
+		NegativeBuckets: []float64{1, 1, 4, 4},
+	}
+
+	f2 := &FloatHistogram{
+		ZeroThreshold:   0.01,
+		ZeroCount:       11,
+		Count:           30,
+		Sum:             2.345,
+		Schema:          1,
+		PositiveSpans:   []Span{{-4, 3}, {5, 5}},
+		PositiveBuckets: []float64{1, 0, 0, 3, 2, 2, 3, 4},
+		NegativeSpans:   []Span{{6, 3}, {6, 4}},
+		NegativeBuckets: []float64{3, 0.5, 0.5, 2, 3, 2, 4},
+	}
+
+	for n := 0; n < b.N; n++ {
+		f1.Add(f2)
```

Comment:
I think that's not a good benchmark as `f1` will have the right schema and the merged layout after the first iteration. It has to be something like `f1.Copy().Add(f2)`.

This would also benchmark the `Copy` method, unfortunately. You could do something more elaborate to avoid that, like creating `b.N` of randomly shaped histograms, then call `b.ResetTimer()` and then add them all up. But let's not overdo it…

Reply:
I took your advice to create randomly shaped histograms which contain 100 spans for each floathistogram in each iteration.(I use `b.StopTimer()` and `b.StartTimer()` to ignore the time took by creating histograms). 

The result shows a much better enhancement on my laptop, nearly 60%
```
BenchmarkAddOld-8   	    3787	    299805 ns/op
BenchmarkAddNew-8   	    8635	    129142 ns/op
```